File: java-devildyw/src/main/java/top/dyw/leetcode/medium/hash/L12整数转罗马数字.java
package top.dyw.leetcode.medium.hash;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Set;

/**
 * @author Devildyw
 * @date 2023/08/03 12:00
 **/
@SuppressWarnings("all")
public class L12整数转罗马数字 {

    public static void main(String[] args) {

    }


    int[] values = new int[]{1000,900,500,400,100,90,50,40,10,9,5,4,1};
    String[] symbols = new String[]{"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};

    public String intToRoman(int num) {
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            //将罗马数字对应字符的值举出在从大到小去遍历 当前值大于等于某一罗马数字时记录该字符，数字并减去相应的值 当数字为0时跳出
            String symbol = symbols[i];
            while (num>=values[i]){
                num-=values[i];
                res.append(symbol);
            }

            if (num==0){
                break;
            }
        }

        return res.toString();

        


    }
}

File: golang-devildyw/middle/L12整数转罗马数字.go
package main

func main() {

}

func intToRoman(num int) string {
	// 方法1: 贪心算法 - 使用切片保持顺序
	// 关键：把所有符号（包括特殊的减法形式）从大到小排列
	var (
		values  = []int{1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1}
		symbols = []string{"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"}
		result  = ""
	)

	for i := 0; i < len(values); i++ {
		count := num / values[i] // 看看可以用当前罗马数字几次
		if count > 0 {
			num -= count * values[i]
			for count > 0 {
				result += symbols[i]
				count--
			}
		}
		if num == 0 {
			return result
		}
	}
	return result
}
